flexmeasures/data/services/time_series.py

```diff
@@ -315,23 +315,63 @@ def set_bdf_source(bdf: tb.BeliefsDataFrame, source_name: str) -> tb.BeliefsData
 def drop_unchanged_beliefs(bdf: tb.BeliefsDataFrame) -> tb.BeliefsDataFrame:
     """Drop beliefs that are already stored in the database with an earlier belief time.
 
+    Also drop beliefs that are already in the data with an earlier belief time.
+
     Quite useful function to prevent cluttering up your database with beliefs that remain unchanged over time.
-    Only works on BeliefsDataFrames with a unique belief time and unique source.
     """
     if bdf.empty:
         return bdf
-    if len(bdf.lineage.belief_times) > 1:
-        raise NotImplementedError("Beliefs should share a unique belief time.")
-    if len(bdf.lineage.sources) > 1:
-        raise NotImplementedError("Beliefs should share a unique source.")
+
+    # Save the oldest ex-post beliefs explicitly, even if they do not deviate from the most recent ex-ante beliefs
+    ex_ante_bdf = bdf[bdf.belief_horizons > timedelta(0)]
+    ex_post_bdf = bdf[bdf.belief_horizons <= timedelta(0)]
+    if not ex_ante_bdf.empty and not ex_post_bdf.empty:
+        # Recursive function call
```

Comment:
This led me to imagine that more than one recursive  level is possible.

Rather, say that we treat each part separately, to avoid the ex-post knowledge (measurements which were forecasted exactly correct) would be lost.